server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java

```diff
@@ -52,17 +61,79 @@ public Port getPort() {
+                                    routes.post(endpoint.getPath(), (req, res) -> next.getAction().apply(req, res));
+                                    break;
+                                case DELETE:
+                                    routes.delete(endpoint.getPath(), (req, res) -> next.getAction().apply(req, res));
+                                    break;
+                                case OPTIONS:
+                                    routes.options(endpoint.getPath(), (req, res) -> next.getAction().apply(req, res));
+                                    break;
+                            }
+                        } else if (route.size() == 2) {
+                            ImmutableList<JMAPRoute> sorted = route.stream()
+                                .sorted(Comparator.comparing(JMAPRoute::getVersion))
+                                .collect(Guavate.toImmutableList());
+                            JMAPRoute draftRoute = sorted.get(0);
+                            JMAPRoute rfc8621Route = sorted.get(1);
+
+                            switch (endpoint.getVerb()) {
+                                case GET:
+                                    routes.get(endpoint.getPath(), (req, res) -> chooseVersionRoute(req, res, draftRoute, rfc8621Route));
+                                    break;
+                                case POST:
+                                    routes.post(endpoint.getPath(), (req, res) -> chooseVersionRoute(req, res, draftRoute, rfc8621Route));
+                                    break;
+                                case DELETE:
+                                    routes.delete(endpoint.getPath(), (req, res) -> chooseVersionRoute(req, res, draftRoute, rfc8621Route));
+                                    break;
+                                case OPTIONS:
+                                    routes.options(endpoint.getPath(), (req, res) -> chooseVersionRoute(req, res, draftRoute, rfc8621Route));
+                                    break;
+                            }
+                        }
+                    }))
+                )
                 .wiretap(configuration.wiretapEnabled())
                 .bindNow());
         }
     }
 
+    private Publisher<Void> chooseVersionRoute(HttpServerRequest request, HttpServerResponse response, JMAPRoute draftRoute, JMAPRoute rfc8621Route) {
+        if (hasRfc8621AcceptHeader(request)) {
+            return rfc8621Route.getAction().apply(request, response);
+        }
+        return draftRoute.getAction().apply(request, response);
+    }
+
+    private boolean hasRfc8621AcceptHeader(HttpServerRequest request) {
+        Optional<String> version = extractRequestVersionHeader(request);
+
+        return JMAPVersion.of(version).equals(JMAPVersion.RFC8621);
+    }
+
+    private Optional<String> extractRequestVersionHeader(HttpServerRequest request) {
+        return Optional.ofNullable(request.requestHeaders().get(JMAP_VERSION_HEADER));
```

Comment:
I'm not sure this is right...

Reply:
No. We should rely on the `Accept` header.

Read the related ADR again